src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -72,4 +109,228 @@ object RuleUtils {
+   *
+   * @param spark Spark session
+   * @param index index used in replacement plan
+   * @param plan current plan
+   * @return replacement plan
+   */
+  def getReplacementPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    // Here we can't replace the plan completely with the index. This will create problems.
+    // For e.g. if Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
+    // if we replace this plan with index Scan (A,B,C), we lose the Filter(C=10) and it will
+    // lead to wrong results. So we only replace the base relation.
+    val hybridScanEnabled = spark.sessionState.conf
+      .getConfString(
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED,
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED_DEFAULT.toString)
+      .toBoolean
+
+    if (hybridScanEnabled) {
+      getHybridScanIndexPlan(spark, index, plan, useBucketSpec)
+    } else {
+      getIndexPlan(spark, index, plan, useBucketSpec)
+    }
+  }
+
+  private def getIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    plan transformUp {
+      case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
+        val location =
+          new InMemoryFileIndex(spark, Seq(new Path(index.content.root)), Map(), None)
+        val relation = HadoopFsRelation(
+          location,
+          new StructType(),
+          StructType(index.schema.filter(baseRelation.schema.contains(_))),
+          if (useBucketSpec) Some(index.bucketSpec) else None,
+          new ParquetFileFormat,
+          Map())(spark)
+
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+  }
+
+  private def getHybridScanIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    var useBucketUnion = useBucketSpec
+    val replacedPlan = plan transformUp {
+      case baseRelation @ LogicalRelation(
+            fsRelation @ HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+            baseOutput,
+            _,
+            _) =>
+        val curFileSet = location.allFiles
+          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+          .toSet
+        // if BucketSpec of index data isn't used, we could read appended data from
+        // source files directly.
+        val filesAppended = if (!useBucketUnion) {
+          (curFileSet -- index.allSourceFiles).map(f => new Path(f.name)).toSeq
+        } else Seq()
+        val filesDeleted =
+          (index.allSourceFiles -- curFileSet).toSeq.map(f =>
+            f.name.replace("file:/C:", "file:///C:")) // TODO fix
```

Comment:
@pirz @apoorvedave1 Please let me know if you have any idea or comment about this. I'd like to change `file:/C:` to `file:///C:`for all platforms as the path in the lineage column starts with `file:///`. Would it be safe to replace `file:/` with `file:///` ?

Reply:
When I create a new Path("C:/a/b/c") to its `absolutePath`, it gets converted to `file:/C:/` automatically, and that's what will get saved in the metadata as per this pr #139 . Is it possible to update the lineage column from `file:///C:` to `file:/C:`? If that's possible, we wouldn't have to do this replace.

cc @pirz

Reply:
Assuming scheme is `file`, we could do something like:
```
val original = "file:/tmp"
val p = new Path(new Path(original).toUri.getPath)
val fs = p.getFileSystem(new Configuration)
p.makeQualified(fs).toUri.toString
```

If we want to remove it when we create the lineage path, we can do `new Path("file:///tmp").toString`.